### DISCUSSION TITLE
What happened to the nextgen engine?

### DISCUSSION THREAD

**Discussion by btserket** [graphql-java/graphql-java#3009] [General]
I've noticed the [nextgen engine](https://github.com/graphql-java/graphql-java/commit/202d443f49bcaa0bfad6278fc62563d15fe94dd1) is gone from the master. I can imagine it wasn't widely used, but it offered an infinitely easier alternative to batching (as compared to DataLoader), worked just fine in many use cases, and... well... was supposed to be the _next generation_.

Maybe @dondonz or @bbakerman can shed some light on the decision.

**Comment by bbakerman** [comment]
The next gen executable strategy was not without merit - especially around how it batched.

However it created other problems

* `Instrumentation` is intimately tied to how the engine executes the fields.   If different strategy implies a different `Instrumentation` and we were not set up for that
* NextGen did lead us to invent structures for representing how to execute a query but the ones used are not the ones we were happy with.  We learnty from it and our efforts in Nadel and `graphql.normalized.ExecutableNormalizedOperation / graphql.normalized.ExecutableNormalizedField` is much more likely  to be the basis for a new engine.
* if we had our time again - we would have NEVER made `ExecutionStrategy` a public API class - rather have some `interface ExecutionStrategy` and the leave it up to others to make new ones based on common starting points.  The reason is that inheritance is a terrible model of code sharing and backwards compatibility on engines has proven.... difficult
* There is a maintenance challenge in having more code

**Comment by andimarek** [comment]
Next gen was pretty much a playground of ideas, but they were never viable enough for a general purpose replacement of the current engine. 

Additionally to points by @bbakerman I would highlight the immutable result tree (`ExecutionResultNode`): it is very helpful if you want to post-process the result, but for a general purpose engine it is simply not fast enough. A full traversal of `ExecutionResultNodes` is needed to produce an actual `ExecutionResult.` 

I agree @btserket that the batched execution is very nice: our goal is still to provide such a batched execution algorithm without dataloader, but we have not really found the right overall design to offer that.

**Comment by btserket** [comment]
Thank you @bbakerman and @andimarek for the explanation! Much appreciated.
